### ISSUE TITLE
Vision: generation degenerates to repeated tokens after successful image encode (embedded skippy path)

### ISSUE THREAD

**Issue by michaelneale** [michaelneale/decentralized-inference#1314]
## Summary

With the mtmd FFI fix from #1313 applied, image input now tokenizes and encodes correctly through the OpenAI surface — but the subsequent generation is garbage: the model emits an endless run of the same token, which the peg-native chat parser then rejects:

```
> common_chat_peg_parse: unparsed peg-native output: e0e0e0e0e0e0e0e0e0e0e0e0e0e0e0e0...
{"error":{"message":"Unsupported: failed to parse chat response: The model produced output that does not match the expected peg-native format", ...}}  (HTTP 502)
```

Native log confirms the vision encode itself succeeds:

```
encoding image slice...
image slice encoded in 566 ms
```

## Control — model and vendored llama.cpp are fine

`llama-mtmd-cli --jinja` built from the **same vendored checkout** (`.deps/llama.cpp`), same GGUF + mmproj (Muse-Glimmer-30B Q4_K_M), same image, describes the test image perfectly (shapes, colors, text all correct). So the break is in the embedded skippy serving path, not the model, mmproj, or mtmd itself.

## Repro

1. `mesh-llm serve --gguf <glimmer.gguf> --mmproj <mmproj.gguf> --port 9447` (main + #1313, M4/Metal)
2. `POST /v1/chat/completions` with content parts `[{type: image_url, image_url: {url: data:image/png;base64,...}}, {type: text, ...}]`
3. 502 with the peg-native parse error; native log shows successful slice encode followed by degenerate output.

Text-only requests on the same serve are unaffected (correct answers, correct tool calls).

## Notes / leads

- Repro'd identically with the prefix cache disabled (`prompt_cache = false`), so it's not cache/lane interaction with the resident prefix tier.
- Glimmer is `MTMD_POS_TYPE_NORMAL` (standard rope), so mrope position handling shouldn't be in play.
- `skippy_runtime::media::eval_media` drives `mtmd_helper_eval_chunks` inside the `begin/end_external_decode` guard with hardcoded `seq_id = 0`; the session's lane comes from `skippy_claim_execution_lane` (0 for the first session, so likely fine single-user — but worth confirming the guard correctly hands logits state back to `skippy_session_sample_current` after an external `llama_decode`).
- The degenerate single-token loop smells like the first sampled token after the external decode reads stale/garbage logits, then the feedback loop repeats it.

Happy to test candidate fixes on this setup — the full stack (Glimmer + mmproj + M4 Metal + repro request bodies) is standing by.

**Comment by michaelneale** [comment]
🤖 Update: closed #1313 rather than landing a partial fix — the FFI correction is a prerequisite for this issue but doesn't make vision work on its own. Whoever picks this up: branch `fix/mtmd-bitmap-helper-ffi` (commit 0da99cf4) contains the reviewed, verified first layer:

- `skippy-ffi` declared the old 3-arg `mtmd_helper_bitmap_init_from_buf`; vendored llama.cpp now takes a 4th `placeholder` bool and returns `{bitmap, video_ctx}`. On aarch64 the bitmap pointer accidentally survives in x0 but `placeholder` reads garbage — nonzero → empty placeholder bitmap → every vision request 502s at the encode gate.
- With that fixed, image slices encode successfully (`image slice encoded in 566 ms`) and then generation degenerates as described above.

So the full repro sequence for this issue **requires** that branch (or equivalent) first — on current main you'll hit the placeholder failure before you can reproduce the sampling bug. Suggested landing order: cherry-pick 0da99cf4, then debug the sampling handoff, land as one PR that makes vision actually work.

**Comment by michaelneale** [comment]
🤖 Further triage — the original 'sampling handoff' lead is **wrong**, updating the diagnosis:

Ruled out (all on branch `fix/mtmd-bitmap-helper-ffi`, Glimmer-30B + mmproj, M4 Metal):
- **Session/lane reuse**: fresh serve, vision as the very first request → still fails (first session claims lane 0, matching the hardcoded `seq_id=0` in `eval_media`).
- **Sampling config**: `temperature: 0` (greedy) → still fails.
- **Prefix cache**: `prompt_cache = false` → still fails (reported earlier).

Key new evidence: under greedy sampling the output is **not** a degenerate token loop — it's coherent, fluent English that is completely unrelated to the image (e.g. rambling about `if (a == b)` comparisons). The peg-native parse failure is downstream: the model just never emits the ATEM protocol structure because it's continuing from noise.

So sampling is fine and generation is fine — the **context the decoder sees after `mtmd_helper_eval_chunks` is garbage**. Suspect area: the external-decode handoff in `skippy_runtime::media::eval_media` — image embedding batch layout / Metal buffer sync / position bookkeeping between mtmd's internal `llama_decode` calls and skippy's session state. Control case still holds: `llama-mtmd-cli --jinja` from the same vendored checkout works perfectly, so the bug is in how the embedded path drives the eval, not in mtmd itself.

**Comment by ndizazzo** [comment]
Confirmed directly on current main a6b16e62220a4ae98849959f5f526662906e7b47 and closing.

Proof:
- Built the release host and packaged Metal runtime with just release-host-build and just release-runtime-build metal.
- Hardware/backend: Apple M4 Pro, MTL0.
- Exact repro artifacts: meta-models/Muse-Glimmer-30B-GGUF at revision 70bf1b61ac09f91b24d39038091b41c582bc5d7a, using Muse-Glimmer-30B-KQuant-17GB-Q4_K_M.gguf plus mmproj-Muse-Glimmer-30B-Q4_K_M.gguf.
- Fresh server, first inference: POST /v1/chat/completions, temperature 0, max_tokens 128, checked-in multimodal-smoke.png. Result was HTTP 200, finish_reason stop, with grounded content: "The central block is filled with a solid bright red."
- Native proof: "encoding image slice..." followed by "image slice encoded in 772 ms"; no unparsed peg-native output, fatal, or panic in that fresh run.
- Text-only control on the same process: 17 multiplied by 23 returned HTTP 200, finish_reason stop, content "391".

The fix is cumulative: #1549 corrected the mtmd by-value struct layouts and #1550 corrected the current bitmap-wrapper signature. #1550 reported that the deeper failure no longer reproduced; this run confirms that on the exact Muse Glimmer plus M4/Metal setup from this issue.

A follow-up request after deliberately ending an earlier multimodal request with finish_reason length still failed. That lifecycle case remains separate under #1551 and is not being treated as part of this closure.
